sciris/sc_parallel.py

```diff
@@ -0,0 +1,260 @@
+        
+        results = parallelize(func=f, iterarg=[1,2,3], ncpus=3)
+    
+    Example 2:
+        import pylab as pl
+        def myfunc(x, y, i):
+            xy = [x+i*pl.rand(1000), y+i*pl.randn(1000)]
+            return xy
+        n = 5
+        xy = parallelize(myfunc, kwargs={'x':3, 'y':8}, iterarg=n, itervar='i', maxload=0.8)
+        for i in range(n):
+            pl.scatter(xy[i][0], xy[i][1])
+    
+    Version: 2018nov01
+    '''
+    # Handle maxload
+    if ncpus is None and maxload is None:
+        ncpus = mp.cpu_count()
+    
+    # Handle iterarg
+    if not(ut.isiterable(iterarg)):
+        try:
+            iterarg = np.arange(iterarg)
+        except Exception as E:
+            errormsg = 'Could not understand iterarg "%s": not iterable and not an integer: %s' % (iterarg, str(E))
+            raise Exception(errormsg)
+    
+    # Construct argument list
+    argslist = []
+    niters = len(iterarg)
+    for index,iterval in enumerate(iterarg):
+        taskargs = TaskArgs(func, index, iterval, itervar, args, kwargs, maxload)
+        argslist.append(taskargs)
+        
+    # Decide how to run -- not load-balanced, use map
+    if ncpus:
+        multipool = mp.Pool(processes=ncpus)
+        outputlist = multipool.map(parallel_task, argslist)
+        return outputlist
+    
+    # It is load-balanced, use Process/Queue
+    else:
+        outputqueue = mp.Queue()
+        outputlist = np.empty(niters, dtype=object)
+        processes = []
+        for i in range(niters):
+            prc = mp.Process(target=parallel_task, args=(argslist[i], outputqueue))
+            prc.start()
+            processes.append(prc)
+        for i in range(niters):
+            _i,returnval = outputqueue.get()
+            outputlist[_i] = returnval
+        for prc in processes:
+            prc.join() # Wait for them to finish
+        
+        outputlist = outputlist.tolist()
+        
+        return outputlist
+
+
+
+def parallel_task(taskargs):
+    ''' Task called by parallelize() -- not to be called directly '''
+    
+    # Handle inputs
+    func = taskargs.func
+    index = taskargs.index
+    args = taskargs.args
+    kwargs = taskargs.kwargs
+    if args   is None: args   = ()
+    if kwargs is None: kwargs = {}
+    if taskargs.itervar is None:
+        args = (taskargs.iterval,) + args # If variable name is not supplied, prepend it to args
+    else: 
+        kwargs[taskargs.itervar] = taskargs.iterval # Otherwise, include it in kwargs
+    
+    # Handle load balancing
+    maxload = taskargs.maxload
+    if maxload:
+        loadbalancer(maxload=maxload, index=index)
+    
+    # Call the function
+    output = func(*args, **kwargs)
+        
+    # Handle output
+    outputqueue = None
+    if outputqueue:
+        outputqueue.put((index,output))
+        return None
+    else:
+        return output
+    
+    
+
+
+def parallelcmd(cmd=None, parfor=None, returnval=None, maxload=None, **kwargs):
```

Comment:
I find `parallelcmd` quite confusing to understand, maybe a bit hacky for production...? I'd consider removal (but up to you). I think it's probably not too great a restriction to have users go via `parallelize` instead, I'd just be concerned about not really knowing what side effects `parallelcmd` may or may not have

Reply:
Agree, it's more just a curiosity :) But I thought it might be useful in some cases (mostly for quick prototyping while needing to make minimal changes to the code). I've added a warning to the docstring to not use it in production code.